src/main/java/org/apache/sysds/runtime/instructions/fed/CovarianceFEDInstruction.java

```diff
@@ -0,0 +1,253 @@
+			long varID = FederationUtils.getNextFedDataID();
+
+			fedMapping.mapParallel(varID, (range, data) -> {
+
+				FederatedResponse response;
+				try {
+					if(input3 == null) {
+						response = data.executeFederatedOperation(new FederatedRequest(FederatedRequest.RequestType.EXEC_UDF,
+							-1,
+							new CovarianceFEDInstruction.COVFunction(data.getVarID(),
+								mb.slice(range.getBeginDimsInt()[0], range.getEndDimsInt()[0] - 1),
+								cop))).get();
+					}
+					else {
+						MatrixBlock wtBlock = ec.getMatrixInput(input2.getName());
+
+						response = data.executeFederatedOperation(new FederatedRequest(FederatedRequest.RequestType.EXEC_UDF,
+							-1,
+							new CovarianceFEDInstruction.COVWeightsFunction(data.getVarID(),
+								mb.slice(range.getBeginDimsInt()[0], range.getEndDimsInt()[0] - 1),
+								cop,
+								wtBlock))).get();
+					}
+
+					if(!response.isSuccessful())
+						response.throwExceptionFromResponse();
+					synchronized(globalCmobj) {
+						globalCmobj.add((CM_COV_Object) response.getData()[0]);
+					}
+				}
+				catch(Exception e) {
+					throw new DMLRuntimeException(e);
+				}
+				return null;
+			});
+
+			Optional<CM_COV_Object> res = globalCmobj.stream().reduce((arg0, arg1) -> (CM_COV_Object) cop.fn.execute(arg0, arg1));
+			try {
+				ec.setScalarOutput(output.getName(), new DoubleObject(res.get().getRequiredResult(cop)));
+			}
+			catch(Exception e) {
+				throw new DMLRuntimeException(e);
+			}
+		}
+	}
+
+	private static class COVFunction extends FederatedUDF {
+
+		private static final long serialVersionUID = -501036588060113499L;
+		private final MatrixBlock _mo2;
+		private final COVOperator _op;
+
+		public COVFunction (long input, MatrixBlock mo2, COVOperator op) {
+			super(new long[] {input});
+			_op = op;
+			_mo2 = mo2;
+		}
+
+		@Override
+		public FederatedResponse execute(ExecutionContext ec, Data... data) {
+			MatrixBlock mb = ((MatrixObject) data[0]).acquireReadAndRelease();
+			return new FederatedResponse(FederatedResponse.ResponseType.SUCCESS, mb.covOperations(_op, _mo2));
+		}
+	}
+
+	private static class COVWeightsFunction extends FederatedUDF {
+		private static final long serialVersionUID = -1768739786192949573L;
+		private final COVOperator _op;
+		private final MatrixBlock _mo2;
+		private final MatrixBlock _weights;
+
+		protected COVWeightsFunction(long input, MatrixBlock mo2, COVOperator op, MatrixBlock weights) {
+			super(new long[] {input});
+			_mo2 = mo2;
+			_op = op;
+			_weights = weights;
+		}
+
+		@Override
+		public FederatedResponse execute(ExecutionContext ec, Data... data) {
+			MatrixBlock mb = ((MatrixObject) data[0]).acquireReadAndRelease();
+			return new FederatedResponse(FederatedResponse.ResponseType.SUCCESS, mb.covOperations(_op, _mo2, _weights));
+		}
+	}
+
+	private static double aggCov(Future<FederatedResponse>[] ffr, Future<FederatedResponse>[] mean1Ffr, Future<FederatedResponse>[] mean2Ffr, FederationMap map) {
```

Comment:
Please split this method up. into at least:
1. that gets the federated responses.
2. that calculates the covariance (using only double inputs.)